feat: Add global job timeout

bgins opened the pull request:
### Summary

This pull request makes the following changes:

- [x] Add expired job cancellation to the solver control loop
  - [x] Add `JOB_TIMEOUT_SECONDS` config
  - [x] Add `JobTimedOut` terminal agreement state
  - [x] Add cancelled job metrics 
  - [x] Add cancelled job traces
- [x] Add `Active` job offer query filter
- [x] Add `Cancelled` job offer query filter
  - [x] Remove `IncludeCancelled` filter and integrate it into `Cancelled` 

We have also made a couple of related changes and small cleanups:

- [x] Rename deal stats metrics to job stats metrics
- [x] Record job stat metrics for cancelled job offers that are not involved in a deal
- [x] Clean up allowed resource providers between test runs. We missed it while rebasing #481.
- [x] Update comments in solver `solve` function
- [x] Add `BoolPointer` language helper to more easily construct pointers to bools

The primary goal of this pull request is adding a global job timeout for jobs that are stuck, stale, or abandoned. The global timeout will be superseded by timeouts in module specs and job offers when we have implemented them.

We have also added new `Active` and `Cancelled` job offer query filters in support of the global timeout.

### Task/Issue reference

Closes: #511

### Test plan

Start the base services. Start the solver with the job offer timeout set to ten seconds:

```
./stack solver --job-timeout-seconds 10
```

Submit a job without a resource provider running. The job should fail and report a job timed out error message.

Start a resource provider and run another job. Again, the job should be cancelled, but this time in flight. The job creator should see the error message and the resource provider should repost their resource offer.

See the details below if you would like to confirm the activity through metrics and traces.

### Details

**`Active` and `Cancelled` job offer filter queries**

We have added two new job offer filter queries. 

When `Active` is `true`, only jobs in `DealNegotiating`, `DealAgreed`, or `ResultsSubmitted` states are included.

Query active job offers from the solver server with the `active` querstring param. In local development:

```sh
curl http://localhost:8081/api/v1/job_offers?active=true
```

The `Cancelled` query filter is a `*bool` where:

- `nil` includes all jobs
- `true` includes jobs in `JobOfferCancelled` and `JobTimedOut` states
- `false` includes jobs _not_ in a `JobOfferCancelled` or `JobTimedOut` state

We removed the `IncludeCancelled` filter query which is now performed when `Cancelled` is `nil`.

Query jobs on the solver server using the `cancelled` querystring param. In local development:

```sh
curl http://localhost:8081/api/v1/job_offers                    # all jobs
curl http://localhost:8081/api/v1/job_offers?cancelled=true     # cancelled jobs   
curl http://localhost:8081/api/v1/job_offers?cancelled=false    # not cancelled jobs
```

**`JobTimedOut` agreement state**

We have added a new `JobTimedOut` agreement state to describe jobs that were cancelled because they expired. 

We considered using the existing `JobOfferCancelled` state, but it is not coherent when we want to mark resource offers and deals with an expired state. In addition, our mechanism for reporting what happened to the job creator is limited, so `JobTimedOut` serves as a tag to help us log a good error message for the job creator.

When a job times out before a deal was made, the job offer is marked with the `JobTimedOut` state. When a deal has already been made, the job offer, resource offer, and deal are all marked with the `JobTimedOut` state.

If a resource provider attempts to upload files or report results for an expired job, the solver server rejects their post with an error reporting the job as timed out.

**`JOB_TIMEOUT_SECONDS` config**

We have added a new config to set the global timeout in seconds. The default value is ten minutes.

Set the job timeout with an environment variable of CLI option. For example, to expire jobs after 30 seconds:

```
./stack solver --job-timeout-seconds 30
JOB_TIMEOUT_SECONDS=30 ./stack solver
```

**Observability**

The deal metrics have been updated to reflect overall job metrics. We want to include metrics about job offers that were never in a deal, and we record these from jobs that expired or were cancelled.

The metrics are reported in an updated `Solver Metrics` dashboard:

![timed-out-metrics](https://github.com/user-attachments/assets/bbeb3590-b9f3-4248-bdd1-d2732ac80efb)

The updated dashboard is available in this observability pull request: https://github.com/Lilypad-Tech/observability/pull/27

Metrics are not enabled by default in local development, but they can be turned on with:

```sh
export ENABLE_METRICS=true
export METRICS_URL="http://localhost:8500"
export METRICS_TOKEN="eyJhbGciOiJFZERTQSIsInR5cCI6IkpXVCJ9.eyJhdXRob3JpemVkIjp0cnVlLCJ1c2VyIjoicmVzb3VyY2UtcHJvdmlkZXIifQ.n36M_ngwC4XPQ_pEkkWAnPiOinnx6-0VO1v_WgCTUEERD7b_p9KHCU6SY5bUdFh5UXRZHAhc1gfyc7rjAnmeDQ"
  ```

We have also included traces over our new `cancelExpiredJobs` solver controller function. Traces are also disabled by default in local development, but can be enabled with the `--disable-telemetry=false` CLI option.

The traces can be queried by searching TraceQL with the `cancel_expired_jobs` span name and a non-empty expired jobs span attribute:

```
{name="cancel_expired_jobs" && span.expired_job_offers!="[]"}
```

When a job is cancelled before a match, the trace will look like:

![cancelled-before-match-trace](https://github.com/user-attachments/assets/40270442-c63b-4c3e-91c8-548fe6a0d779)

If a job is cancelled after a match, it will also report a deal ID:

![cancelled-post-match](https://github.com/user-attachments/assets/b5a8be2b-d0e6-46b0-986e-6453f3fa61ec)

### Related issues or PRs

Observability dashboard PR: https://github.com/Lilypad-Tech/observability/pull/27

bgins committed:
test: Clean up allowed resource providers between test runs

bgins committed:
feat: Add active job offers filter query

bgins committed:
feat: Add job offer timeout field and options

bgins committed:
feat: Add terminal JobTimedOut agreement state

bgins committed:
chore: Rename solver dealStats to jobStats

bgins committed:
feat: Add cancelled job offers filter query

bgins committed:
feat: Add cancelled job metrics

bgins committed:
feat: Add cancelExpiredJobs

bgins committed:
chore: Update comments

bgins committed:
chore: Rename reportDealMetrics to reportJobMetrics

narbs91 commented on pkg/options/solver.go:
```diff
@@ -26,6 +29,10 @@ func AddSolverCliFlags(cmd *cobra.Command, options *solver.SolverOptions) {
 	AddTelemetryCliFlags(cmd, &options.Telemetry)
 	AddMetricsCliFlags(cmd, &options.Metrics)
+	cmd.PersistentFlags().IntVar(
+		&options.JobOfferTimeout, "job-offer-timeout", options.JobOfferTimeout,
```
Can we change `job-offer-timeout` to `job-offer-timeout-in-seconds` so that it's clear what unit of measure should be set?

bgins replied:
It's a bit verbose. The help text for the option says it:

https://github.com/Lilypad-Tech/lilypad/blob/bcf3160d901e52869cc95f0ccf6e0e31d8a30925/pkg/options/solver.go#L34

That seem like enough to communicate it?

narbs91 replied:
i personally always like to be precise with this types of things but don't mind if we want to just leave the text to be the reference.  Will defer to you

bgins replied:
We discussed this a bit more and decided that there are places where we don't have the CLI help text to provide context. We've updated this to `JOB_TIMEOUT_SECONDS` to add the unit and to communicate that we are expiring not just offers, but jobs.

narbs91 commented on pkg/options/solver.go:
```diff
@@ -49,6 +56,9 @@ func CheckSolverOptions(options solver.SolverOptions) error {
 	if err != nil {
 		return err
 	}
+	if options.JobOfferTimeout <= 0 {
```
Did we want to enforce a minimum timeout at this point if someone sets it?  Like someone could technically set a timeout of a second if they wanted to (but obviously probably wouldn't).  Maybe jus the 600s you put in the default?

bgins replied:
There are legitimate use cases for a very low timeout. For example, testing this PR we want to set a low timeout for testing. Can see how it could be useful for other local development scenarios and future integration tests.

narbs91 replied:
Ah good point, makes sense to me

narbs91 commented on pkg/options/solver.go:
```diff
@@ -1,19 +1,22 @@
+		Services:        GetDefaultServicesOptions(),
+		Telemetry:       GetDefaultTelemetryOptions(),
+		Metrics:         GetDefaultMetricsOptions(),
+		JobOfferTimeout: GetDefaultServeOptionInt("JOB_OFFER_TIMEOUT", 600),
```
[nit] Maybe have the 600 as a constant variable so that it can be easily set/referenced if needed?

bgins replied:
Would be open to this change if we have a specific use case in mind. Could see readability as a potential motivation, but it would be an exception to how we generally set the defaults for configs.

Perhaps we could take it up separately and think about what patterns we want for our options?

narbs91 replied:
The use case I was thinking of was if we wanted to enforce this a minimum timeout to set but if we're not doing that, ok to leave it as is

narbs91 commented on pkg/options/solver.go:
```diff
@@ -1,19 +1,22 @@
+		Services:        GetDefaultServicesOptions(),
+		Telemetry:       GetDefaultTelemetryOptions(),
+		Metrics:         GetDefaultMetricsOptions(),
+		JobOfferTimeout: GetDefaultServeOptionInt("JOB_OFFER_TIMEOUT", 600),
```
Maybe we can also rename this `JobOfferTimeoutInSeconds`

bgins replied:
See above, we document in the help text, which may be enough? Perhaps we add a comment here to make it clearer when reading the code.

narbs91 replied:
Yup, comment would do the trick too

bgins replied:
Sounds good! Added a comment. 👌

narbs91 commented on pkg/solver/controller.go:
```diff
@@ -321,6 +336,67 @@ func (controller *SolverController) solve(ctx context.Context) error {
+					controller.log.Error("update expired job offer state failed", err)
+					span.SetStatus(codes.Error, "update expired job offer state failed")
+					span.RecordError(err)
+					return err
```
Do we completely want to break the loop on failure vs continuing?

bgins replied:
Yeah great point! Logging and tracing should be enough here. Will update. 👌

bgins replied:
Done ✅

narbs91 commented on pkg/solver/controller.go:
```diff
@@ -321,6 +336,67 @@ func (controller *SolverController) solve(ctx context.Context) error {
+					controller.log.Error("update expired deal state failed", err)
+					span.SetStatus(codes.Error, "update expired deal state failed")
+					span.RecordError(err)
+					return err
```
Ditto here?

bgins replied:
Removed ✅

narbs91 commented on pkg/solver/solver.go:
```diff
@@ -15,12 +15,13 @@ import (
+	Services        data.ServiceConfig
+	Telemetry       system.TelemetryOptions
+	Metrics         system.MetricsOptions
+	JobOfferTimeout int
```
Same comment here about changing to `JobOfferTimeoutInSeconds`

bgins replied:
Same comment as above 🙂

narbs91 commented on pkg/solver/store/db/db.go:
```diff
@@ -159,8 +159,23 @@ func (store *SolverStoreDatabase) GetJobOffers(query store.GetJobOffersQuery) ([
+			data.GetAgreementStateIndex("JobOfferCancelled"),
+			data.GetAgreementStateIndex("JobTimedOut"),
+		})
+	} else if !query.IncludeCancelled {
```
Does there need to be both a Cancelled and a IncludeCancelled?  Is there a different purpose between the two?

bgins replied:
Yes, the purpose of `IncludeCancelled` is when you want to query all offers _except_ cancelled ones. The new `Cancelled` query is when you want to retrieve _only_ cancelled offers.

bgins replied:
Oh actually, maybe we can repurpose `Cancelled`:

- When `Cancelled` is `true`, only return offers in a `JobOfferCancelled` or a `JobTimedOut` state.
- When `Cancelled` is `false`, return all offers not in a `JobOfferCancelled` or a `JobTimedOut` state.

bgins replied:
Following up, we added a `nil` state where `Cancelled` is not present to query for all job offers.

narbs91 commented on pkg/solver/store/store.go:
```diff
@@ -20,6 +20,13 @@ type GetJobOffersQuery struct {
+	Active bool `json:"in_progress"`
+
+	// Cancelled job offers are in a JobOfferCancelled or JobTimedOut state
+	Cancelled bool `json:"cancelled"`
```
Ah I see, ones for including results the other is for clearing out the stale offers.  Perhaps we can change the name of this to `inactive` or `terminated` so that it doesn't clash with `IncludeCancelled`?

bgins replied:
Wrote some more thoughts about this above, we may be able to simplify it to only use `Cancelled`.

bgins committed:
chore: Remove exits on update state failures

bgins committed:
chore: Add comment about JOB_OFFER_TIMEOUT unit and defaults

bgins committed:
feat: Add BoolPointer lang helper

bgins committed:
feat: Combine cancelled filter queries

The cancelled filter query now includes the include cancelled, which was
our mechanism for requesting all job offers. The cancelled filter query
includes all when nil, or cancelled or not when true or false.

bgins committed:
reafctor: Rename JOB_OFFER_TIMEOUT to JOB_TIMEOUT_SECONDS

bgins committed:
chore: Improve timeout option comment

bgins committed:
feat: Reject results and file uploads for timed out jobs

bgins committed:
chore: Improve results checking log message

narbs91 commented on pkg/solver/store/memory/store.go:
```diff
@@ -105,9 +105,26 @@ func (s *SolverStoreMemory) GetJobOffers(query store.GetJobOffersQuery) ([]data.
 			matching = false
 		}
+		if query.Cancelled != nil {
+			isCancelled := jobOffer.State == data.GetAgreementStateIndex("JobOfferCancelled") ||
```
[nit] maybe change this variable to be `isJobOfferInStaleState` so that it doesn't clash with the `query.Cancelled`

bgins replied:
Updated to improve readability ✅

narbs91 reviewed (approved):
LGTM!

bgins committed:
refactor: Clean up GetJobOffers memory store method logic
